### ISSUE TITLE
Session delete_old_data()  is deleting sessions it should not be with no regard to cookie settings.

### ISSUE THREAD

**Issue by sdoherty0** [ExpressionEngine/ExpressionEngine#522]
**Description of the problem**

The delete_old_data() in the libraries/Session.php class is called whenever a session is constructed. When it is called it deletes 'expired' sessions. However, the code is based on whoever happens to land on that page at a given time. The $this->session_length depends on whether the user is an admin or not. If an admin is logged in and happens to trigger this, then the $expire value will be very small (default one hour), and other users whose last_activity is past that will have their sessions deleted, and that without any reference to their expiration cookie.  If you delete a session entry from the database, even with 'c' (just cookies) setting, the login is lost.

libraries/Session.php
```	
public function delete_old_sessions()
	{
		$expire = ee()->localize->now - $this->session_length;

		srand(time());

		if ((rand() % 100) < $this->gc_probability)
		{
			ee()->db->where('last_activity < ', $expire)
						 ->delete('sessions');
		}
	}
```

**How To Reproduce**
1. Log an account into an expression engine site with cookies only setting for user validation.
2. Open your database software (e.g. phpmyadmin) and find the sessions table.
3. Delete the session entry corresponding to their member_id.
4. Their login will now be expired (thus proving what will be the effect of delete_old_data())
5. Therefore, every time that delete_old_data() removes a session entry, the user will be logged out, regardless of their cookie expiration date, thus defeating cookies.

**Environment Details:**
 - Version: 2.11.6  (but the code is still the same in 5.3.2)
 - PHP Version 5.6.40
 - Web Server:  Apache

**Possible Solution**

- The code should not have the $expire variable dependent on who is hitting the code at any given time (i.e. admin or user) because the non-admins will have their session prematurely expired.
- The code should check if the validation method is cookie only, because then it would need to check against the cookie's expiration rather than now minus session_length.

**Additional context**

It seems that this might be the cause of a lot of other premature log out reports from others here and on the forum.

**Comment by intoeetive** [comment]
@sdoherty0 thank you for the report! It looks like this might be causing the "premature logout" issue indeed. We're investigating this deeper now.

**Comment by sdoherty0** [comment]
I have further found that if the automatically log on in future visits "auto_login" form field is checked that the session will be recreated from the cookie. But if it is not, then it won't. I think this behavior has masked the errant delte_old_data() behavior by preserving those users with the checkbox checked. But that is not the intended function of that checkbox. It really is functioning as a protect_me_from_delete_old_data() flag. If you don't have it checked and your session is older than the (race condition) last person to land on a page and trigger Session.php you will be logged out because your session is destroyed, and there is no entry in the exp_remember_me table that says to recreate your session from your cookie. Therefore the expiration date set for users becomes meaningless.

**Comment by intoeetive** [comment]
This has been fixed in 6.0.0 release
